Added Creditor Name + Several Bugfixes

fhasse95 opened the pull request:
Added Creditor Name
- The creditor name has been added to transactions

Bugfixes
- Fixed the simple date formatter string (DD is day of year; dd is day of month | YYYY is week of year; yyyy is calendar year)
- Fixed a decoding error (no acceptance date) that occurs when creating a new requisition request within the "connectToInstitution" function with a custom agreement
- Fixed a decoding error (no decimal value) that occurs when fetching a monetary transaction amount with certain banks (e.g. "DKB")

fhasse95 committed:
Added Creditor Name + Several Bugfixes

Added Creditor Name
- The creditor name has been added to transactions

Bugfixes
- Fixed the simple date formatter string (DD is day of year; dd is day of month | YYYY is week of year; yyyy is calendar year)
- Fixed a decoding error (no acceptance date) that occurs when creating a new requisition request within the "connectToInstitution" function with a custom agreement
- Fixed a decoding error (no decimal value) that occurs when fetching a monetary transaction amount with certain banks (e.g. "DKB")

Arestronaut commented:
@fhasse95 
I will take a look at the PR asap. Only thing I noticed immediately are the changes to `MonetaryAmount`. I specifically don't want the `decimalValue` to be optional as whenever the struct is initialised with values that cannot be described as a `Decimal` all the operations wouldn't make any sense by just defaulting to zero. So if one Monetary value is invalid the whole operation should not be possible in the first place. WDYT?

fhasse95 committed:
Changed the decimal value to a calculated variable

fhasse95 commented:
@Arestronaut The problem I had was that I got a decoding error when I tried to retrieve certain data (e.g. balance or transactions). This was probably because the response from the server only returns the "amount" and "currency" fields. However, I understand your point that the operations don't make sense if I use zero as fallback. I have now changed the decimal value to a calculated variable so that the operations can still be used but no decoding error occurs.

Arestronaut commented on NordigenSwiftDemo/NordigenSwiftDemo/Scenes/Main/ContentViewModel.swift:
```diff
@@ -98,17 +100,29 @@ final class ContentViewModel: ObservableObject {
         Task {
             do {
-                let requisition = try await nordigenClient.RequisitionsAPI.new( .init(institutionId: sandboxInstitutionId, redirect: "NordigenSwiftDemo://BankAuthenticationRedirect?id=\(sandboxInstitutionId)"))
+                let payload = EndUserAgreementWriteRequest(
```
Creating a custom EndUserAgreement should not be hardcoded for the default use case. I'd suggest to create a separate flow to do so

fhasse95 replied:
I have removed the custom end user agreement from the default use case in the latest commit.

Arestronaut commented on Sources/NordigenSwift/Models/MonetaryAmount.swift:
```diff
@@ -8,28 +8,20 @@ import Foundation
-        self.amount = amount
-        self.currency = currency
-
+    public var decimalValue: Decimal {
```
My initial reasoning why I didn't want this to be a computed property was that this value only needs to be calculated once. But I do see the flaw in the current implementation.
A possible workaround could be: 
```swift
public struct MonetaryAmount: Codable, Equatable, Hashable {
    enum CodingKeys: CodingKey {
        case amount
        case currency
        case decimalValue
    }

    public let amount: String
    public let currency: String

    public private(set) var decimalValue: Decimal!

    public init(amount: String, currency: String) {
        self.amount = amount
        self.currency = currency

        createDecimalValue()
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.amount = try container.decode(String.self, forKey: .amount)
        self.currency = try container.decode(String.self, forKey: .currency)

        createDecimalValue()
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(self.amount, forKey: .amount)
        try container.encode(self.currency, forKey: .currency)
    }

    private mutating func createDecimalValue() {
        if let nsDecimal = NumberFormatters.decimalNumberFormatterDotSeparated.number(from: amount) as? NSDecimalNumber {
            decimalValue = nsDecimal.decimalValue
        } else if let nsDecimal = NumberFormatters.decimalNumberFormatterCommaSeparated.number(from: amount) as? NSDecimalNumber {
            decimalValue = nsDecimal.decimalValue
        } else {
            preconditionFailure("Amount(\(amount)) is not a valid decimal number")
        }
    }
}
```

fhasse95 replied:
I think this will fix the problem with the decoding error. Therefore, I have adopted your suggestion in the latest commit.

fhasse95 committed:
Code Review Comments

- Removed the custom end user agreement from the default use case
- Updated the MonetaryAmount struct to calculate the decimal value only once

fhasse95 committed:
Extended Transaction + Re-Added the formatted monetary amount

Arestronaut commented on Sources/NordigenSwift/Models/MonetaryAmount.swift:
```diff
@@ -6,22 +6,47 @@
+    enum CodingKeys: CodingKey {
+        case amount
+        case currency
+        case decimalValue
```
The CodingKey `decimalValue` is not needed

Arestronaut commented:
> Are there any updates regarding this pull request yet?

@fhasse95 Yes sorry, I had a busy few last days. Let's get this done by eow

fhasse95 committed:
Update MonetaryAmount.swift

Arestronaut reviewed (approved):
LGMT

fhasse95 committed:
Merge branch 'main' into Bugfixes

fhasse95 commented:
Can you merge the PR into the master? Unfortunately I can not do that :)
